## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd
import numpy as np

df = pd.read_csv('Resources/client_dataset.csv')
df.head(20)

first      last                                           job  \
0      Donald   Harding                                  Immunologist   
1     Tiffany     Myers                               Music therapist   
2     Shannon    Watson                                  Immunologist   
3      Nathan     Baker                         Accounting technician   
4   Christina  Schwartz                                  Chiropractor   
5     Jessica     Reyes  Production designer, theatre/television/film   
6       Brian     Moore                                       Barista   
7       Jesus    Wright                            Theme park manager   
8     Rebecca     Brown                                   Chiropodist   
9        Greg    Garcia                            Secretary, company   
10       Kyle  Williams           Engineer, broadcasting (operations)   
11    Roberto    Howard                      Surveyor, land/geomatics   
12    Jessica     Reyes  Production designer, theatre/television/film   
13   Patricia     Gould                                      Lobbyist   
14  Katherine    Taylor                        Psychologist, clinical   
15    Matthew    Orozco                    Research scientist (maths)   
16     Joseph    Jordan            Government social research officer   
17     Carrie     Green                           Colour technologist   
18      David     Reyes              Production assistant, television   
19     Shelby  Williams                           Colour technologist   

                    phone                                     email  \
0      793-904-7725x39308          harding.donald.7185@sullivan.com   
1        201.442.4543x942         myers.t.6537@ferguson-johnson.net   
2       687.737.9424x8503                     swatson8146@payne.net   
3        827-788-8123x012                    bakernathan@benson.com   
4            265-829-3643       christinaschwartz9252@mcconnell.com   
5              5016915923                  jessica.reyes@grimes.net   
6     +1-893-400-2733x982               moore.official@martinez.com   
7      605.947.9467x66513             wright.jesus.7579@meadows.com   
8        665-955-0156x926                      rbrown850@lewis.info   
9      (231)342-7216x1421                       ggarcia@bentley.biz   
10        +1-432-234-2942  communitywilliams9024@taylor-padilla.biz   
11           311.284.4575      roberto.howard.1290@davis-reese.info   
12             5016915923                  jessica.reyes@grimes.net   
13  +1-694-749-5807x94968                  newspapergould@walsh.biz   
14      (943)632-3743x028         taylor.katherine.733@odonnell.org   
15   +1-259-213-7449x4333                    orozcomatthew@carr.com   
16     680.657.3271x16477             jordan_joseph_4684@gibson.net   
17    +1-242-699-5329x474                  c.green@oliver-smith.com   
18           714.475.3974                    davidreyes@daniels.org   
19       664.584.8160x389               williamstry70@armstrong.com   

    client_id  order_id  order_date  order_week  order_year         item_id  \
0       58515   8953482  2023-04-28          17        2023  EUD29711-63-6U   
1       37609   8069089  2023-05-19          20        2023  XDA18116-89-4A   
2       57113   1902144  2023-01-29           4        2023  ABE59463-05-7E   
3       46554   9031802  2023-04-25          17        2023  ZMM00836-65-0C   
4       92089   1322274  2023-05-28          21        2023  BZX55559-12-3X   
5       33615   1327730  2023-05-02          18        2023  BMB23158-88-9A   
6       87213   8080280  2023-04-19          16        2023  UAC50632-25-4E   
7       46709   8946782  2023-02-14           7        2023  UEU25316-01-1U   
8       73362   9394118  2023-05-13          19        2023  DDA06248-98-7Z   
9       58789   9665852  2023-04-14          15        2023  DUU85042-87-5E   
10      81012   5775924  2023-04-12          15        2023  DCD61783-36-6B   
11      74112   2859900  2023-02-0

In [3]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [7]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [5]:
# What three item categories had the most entries?
top_3_categories = df['category'].value_counts().head(3)
top_3_categories = pd.DataFrame(top_3_categories)
top_3_categories

count
category          
consumables  23538
furniture    11915
software      8400

In [6]:
# For the category with the most entries, which subcategory had the most entries?

top_subcategory = df['subcategory'].value_counts().head(1)
top_subcategory = pd.DataFrame(top_subcategory)
top_subcategory

count
subcategory             
bathroom supplies   6424

In [170]:
# Which five clients had the most entries in the data?
#client_df = df['client_id'].value_counts().head(5)

client_df = df['client_id'].value_counts().head(5)
client_df
# client_df = pd.DataFrame(client_df)
# client_df

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [166]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df['client_id'].value_counts().head(5).index.tolist()
print(top_5_clients)

[33615, 66037, 46820, 38378, 24741]


In [9]:
# How many total units (the qty column) did the client with the most entries order order?
filtered_df = df[df['client_id'] == 33615]
top_qty = filtered_df['qty'].sum()
top_qty

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
df[['unit_price' , 'qty' , 'line_subtotal']].head(2)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [54]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
# total_df = df [['unit_price' , 'qty']].copy()
# total_df['line_subtotal'] = total_df['unit_price'] * total_df['qty']
# total_df.head(2)

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']
#if the total weight is > 50 pounds then price = $7, else < 50 pounds then price >= $10
#check np.where documentation
df['price'] = np.where(df['total_weight'] > 50, 7, 10)
df['shipping_price'] = df['total_weight'] * df['price']
df[['unit_price' , 'unit_weight' , 'qty' , 'total_weight' , 'shipping_price']].head()
# print(df.columns)
df = df.drop(columns='price')
# df.columns


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [15]:
df = df.drop(columns='price')

In [16]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax_rate = 1.0925
df['line_price'] = (df['line_subtotal'] + df['shipping_price']) * (sales_tax_rate)
df['line_price'] = df['line_price'].round(2)
df[['line_subtotal' , 'shipping_price' , 'line_price']].head(3)

# # Add a new column for total_price
# sales_tax_rate = 0.0925
# df['total_price'] = (df['shipping_price'] + df['line_subtotal'] * (1 + sales_tax_rate))

# print(df)


line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [17]:
# df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
# df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [18]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [20]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['line_price'] - df['line_cost']
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [22]:
# Check your work using the totals above
# put order ids in a list
client_order_ids = [2742071, 2173913, 6128929]

#define a function
def get_order_total(order_id):
    new_df = df.loc[df['order_id'] == order_id, 'line_price']
    return round(new_df.sum(), 2)

for order_id in client_order_ids:
    print(f"Order {order_id} total: $ {get_order_total(order_id)}")

Order 2742071 total: $ 152811.89
Order 2173913 total: $ 162388.71
Order 6128929 total: $ 923441.25


In [15]:
# Check your work using the totals above
# put order ids in a list

#define a function


Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [28]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
df1 = df[df['client_id'].isin(top_5_clients)]
df1.head()

first     last                                           job  \
5    Jessica    Reyes  Production designer, theatre/television/film   
12   Jessica    Reyes  Production designer, theatre/television/film   
137    Bryan    Myers                                Chief of Staff   
214    Bryan    Myers                                Chief of Staff   
216   Angela  Everett                                     Osteopath   

                phone                     email  client_id  order_id  \
5          5016915923  jessica.reyes@grimes.net      33615   1327730   
12         5016915923  jessica.reyes@grimes.net      33615   6059361   
137  444.233.9146x233     bmyers3696@wilson.com      46820   7109502   
214  444.233.9146x233     bmyers3696@wilson.com      46820   8967498   
216  729-620-6192x875    aeverett@rodriguez.com      66037   1203693   

     order_date  order_week  order_year  ... unit_cost unit_weight  qty  \
5    2023-05-02          18        2023  ...    154.44        2.79   78   
12   2023-05-16          20        2023  ...    198.38       27.51  568   
137  2023-05-12          19        2023  ...    111.47        5.65  118   
214  2023-02-06           6        2023  ...      6.26        7.05  133   
216  2023-04-09          14        2023  ...    166.09        0.26  157   

     line_number  line_subtotal  total_weight  shipping_price  line_price  \
5              0       13481.52        217.62         1523.34    16392.81   
12             2      130492.32      15625.68       109379.76   262060.25   
137            9       16063.34        666.70         4666.90    22647.79   
214            3        1061.34        937.65         6563.55     8330.19   
216            2       47393.59         40.82          408.20    52223.46   

     line_cost  line_profit  
5     13569.66      2823.15  
12   222059.60     40000.65  
137   17820.36      4827.43  
214    7396.13       934.06  
216   26484.33     25739.13  

[5 rows x 24 columns]

In [45]:
df2 = df1.groupby('client_id')[['line_price']].sum().reset_index()
df2


client_id   line_price
0      24741  82268892.02
1      33615   8377308.52
2      38378  12906550.87
3      46820   9743794.36
4      66037  10259514.79

In [57]:
df2['formatted_output'] = df2.apply(
lambda row: f"client_id {row['client_id']}: {row['line_price']}", axis=1)

In [58]:
formatted_outputs = df2['formatted_output'].tolist()
for output in formatted_outputs:
    print(output)

client_id 24741: 82268892.02
client_id 33615: 8377308.52
client_id 38378: 12906550.870000001
client_id 46820: 9743794.36
client_id 66037: 10259514.790000001


In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [59]:
df2.head()

client_id   line_price                     formatted_output
0      24741  82268892.02         client_id 24741: 82268892.02
1      33615   8377308.52          client_id 33615: 8377308.52
2      38378  12906550.87  client_id 38378: 12906550.870000001
3      46820   9743794.36          client_id 46820: 9743794.36
4      66037  10259514.79  client_id 66037: 10259514.790000001

In [180]:
top_5_summary_df = df[df['client_id'].isin(top_5_clients)]
top_5_summary_df

first     last                                           job  \
5        Jessica    Reyes  Production designer, theatre/television/film   
12       Jessica    Reyes  Production designer, theatre/television/film   
137        Bryan    Myers                                Chief of Staff   
214        Bryan    Myers                                Chief of Staff   
216       Angela  Everett                                     Osteopath   
...          ...      ...                                           ...   
54535     Angela  Everett                                     Osteopath   
54536     Kendra  Garrett                        Pension scheme manager   
54579  Alexandra    Young                                     Dietitian   
54589  Alexandra    Young                                     Dietitian   
54613     Kendra  Garrett                        Pension scheme manager   

                  phone                            email  client_id  order_id  \
5            5016915923         jessica.reyes@grimes.net      33615   1327730   
12           5016915923         jessica.reyes@grimes.net      33615   6059361   
137    444.233.9146x233            bmyers3696@wilson.com      46820   7109502   
214    444.233.9146x233            bmyers3696@wilson.com      46820   8967498   
216    729-620-6192x875           aeverett@rodriguez.com      66037   1203693   
...                 ...                              ...        ...       ...   
54535  729-620-6192x875           aeverett@rodriguez.com      66037   7550600   
54536      808.257.2029  garrettkendra8479@jefferson.com      24741   1886506   
54579  265-666-3155x118           ayoung9158@stevens.com      38378   8661147   
54589  265-666-3155x118           ayoung9158@stevens.com      38378   4508675   
54613      808.257.2029  garrettkendra8479@jefferson.com      24741   9114777   

       order_date  order_week  order_year  ... unit_cost unit_weight  qty  \
5      2023-05-02          18        2023  ...    154.44        2.79   78   
12     2023-05-16          20        2023  ...    198.38       27.51  568   
137    2023-05-12          19        2023  ...    111.47        5.65  118   
214    2023-02-06           6        2023  ...      6.26        7.05  133   
216    2023-04-09          14        2023  ...    166.09        0.26  157   
...           ...         ...         ...  ...       ...         ...  ...   
54535  2023-02-24           8        2023  ...     10.66        7.52  108   
54536  2023-05-26          21        2023  ...     40.05        4.70  721   
54579  2023-03-09          10        2023  ...     20.29       13.07  118   
54589  2023-03-25          12        2023  ...     38.26        0.70  685   
54613  2023-02-13           7        2023  ...     59.32       13.13   91   

       line_number  line_subtotal  total_weight  shipping_price  line_price  \
5                0       13481.52        217.62         1523.34    16392.81   
12               2      130492.32      15625.68       109379.76   262060.25   
137              9       16063.34        666.70         4666.90    22647.79   
214              3        1061.34        937.65         6563.55     8330.19   
216              2       47393.59         40.82          408.20    52223.46   
...            ...            ...           ...             ...         ...   
54535            2        1810.08        812.16         5685.12     8188.51   
54536            3       46014.22       3388.70        23720.90    76185.62   
54579            3        3687.50       1542.26        10795.82    15823.03   
54589            0       35784.40        479.50         3356.50    42761.43   
54613            2        8705.97       1194.83         8363.81    18648.73   

       line_cost  line_profit  
5       13569.66      2823.15  
12     222059.60     40000.65  
137     17820.36      4827.43  
214      7396.13       934.06  
216     26484.33     25739.13  
...          ...          ...  
54535    6836.40      1352.11  
54536  

In [181]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

top_client_summary_df = top_5_summary_df.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'line_price': 'sum',
    'line_cost' : 'sum',
    'line_profit' : 'sum'
})

top_client_summary_df


qty  shipping_price   line_price    line_cost  line_profit
client_id                                                               
24741      239862      5126448.37  82268892.02  45688899.71  36579992.31
33615       64313      1828984.89   8377308.52   6175313.91   2201994.61
38378       73667      3429455.40  12906550.87   9634720.98   3271829.89
46820       75768      1601448.84   9743794.36   7007191.64   2736602.72
66037       43018      1395151.85  10259514.79   7004482.98   3255031.81

In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

 #DONE

client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [182]:
top_client_summary_df = top_client_summary_df.reset_index()
top_client_summary_df


client_id     qty  shipping_price   line_price    line_cost  line_profit
0      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
1      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
3      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
4      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81

In [183]:
# Format the data and rename the columns to names suitable for presentation.

top_client_summary_df = top_client_summary_df.rename(columns = {
    'client_id' : 'Client ID' , 
    'qty' : 'Units' , 
    'shipping_price' : 'Shipping (millions)' , 
    'line_price' : 'Total Revenue (millions)',
    'line_cost' : 'Total Cost (millions)' ,
    'line_profit' : 'Total Profit (millions)'
})

top_client_summary_df

# # Define the money columns. 
def currency_format_millions(amount):
    return (amount/1000000)


# # Define a function that converts a dollar amount to millions.
columns_to_convert = ['Shipping (millions)' , 'Total Revenue (millions)', 'Total Cost (millions)', 'Total Profit (millions)']
top_client_summary_df[columns_to_convert] = top_client_summary_df[columns_to_convert].apply(currency_format_millions)
top_client_summary_df

# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862             5.126448                 82.268892   
1      33615   64313             1.828985                  8.377309   
2      38378   73667             3.429455                 12.906551   
3      46820   75768             1.601449                  9.743794   
4      66037   43018             1.395152                 10.259515   

   Total Cost (millions)  Total Profit (millions)  
0              45.688900                36.579992  
1               6.175314                 2.201995  
2               9.634721                 3.271830  
3               7.007192                 2.736603  
4               7.004483                 3.255032

In [187]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
final_client_summary_df = top_client_summary_df.sort_values(by= 'Total Profit (millions)', ascending = False)
final_client_summary_df

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862             5.126448                 82.268892   
2      38378   73667             3.429455                 12.906551   
4      66037   43018             1.395152                 10.259515   
3      46820   75768             1.601449                  9.743794   
1      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
0              45.688900                36.579992  
2               9.634721                 3.271830  
4               7.004483                 3.255032  
3               7.007192                 2.736603  
1               6.175314                 2.201995